packages/type-safe-api/README.md

```diff
@@ -0,0 +1,1423 @@
+import { AwsCdkTypeScriptApp } from "projen/lib/awscdk";
+
+// Create the monorepo
+const monorepo = new NxMonorepoProject({ ... });
+
+// Create the API project
+const api = new TypeSafeApiProject({
+  name: "myapi",
+  parent: monorepo,
+  outdir: 'packages/api',
+  // Smithy as the model language. You can also use ModelLanguage.OPENAPI
+  modelLanguage: ModelLanguage.SMITHY,
+  modelOptions: {
+    smithy: {
+      serviceName: {
+        namespace: 'com.mycompany',
+        serviceName: 'MyApi',
+      },
+    },
+  },
+  // CDK infrastructure in TypeScript
+  infrastructureLanguage: Language.TYPESCRIPT,
+  // Generate clients in TypeScript, Python and Java
+  clientLanguages: [Language.TYPESCRIPT, Language.PYTHON, Language.JAVA],
+  // Generate HTML documentation
+  documentationFormats: [DocumentationFormat.HTML_REDOC],
+});
+
+// Create a CDK infrastructure project
+const infra = new AwsCdkTypeScriptApp({ ... });
+
+// Infrastructure can depend on the generated API infrastructure and client
+infra.addDeps(api.typescriptInfrastructure!.package.packageName);
+infra.addDeps(api.typescriptClient!.package.packageName);
+
+monorepo.synth();
+```
+
+#### Use the CDK Construct
+
+In your CDK application, consume the `Api` construct, vended from the generated typescript infrastructure package.
+
+```ts
+import { Stack, StackProps } from 'aws-cdk-lib';
+import { Construct } from 'constructs';
+import { Api } from "myapi-typescript-infra"; // <- generated typescript infrastructure package
+import { Authorizers, Integrations } from "@aws-prototyping-sdk/type-safe-api";
+import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
+import { Cors } from "aws-cdk-lib/aws-apigateway";
+import * as path from 'path';
+
+export class MyStack extends Stack {
+  constructor(scope: Construct, id: string, props: StackProps = {}) {
+    super(scope, id, props);
+
+    // Instantiate the generated CDK construct to deploy an API Gateway API based on your model
+    new Api(this, 'MyApi', {
+      defaultAuthorizer: Authorizers.iam(),
+      corsOptions: {
+        allowOrigins: Cors.ALL_ORIGINS,
+        allowMethods: Cors.ALL_METHODS,
+      },
+      // Supply an integration for every operation
+      integrations: {
+        sayHello: {
+          integration: Integrations.lambda(new NodejsFunction(this, 'SayHelloLambda', {
+            entry: path.resolve(__dirname, 'say-hello.ts'),
+          })),
+        },
+      },
+    });
+  }
+}
+```
+
+#### Implement a Lambda Handler
+
+The generated client projects include lambda handler wrappers which provide type-safety for implementing your API operations. You can implement your lambda handlers in any of the supported languages, and even mix and match languages for different operations if you like. In typescript, the implementation of `say-hello.ts` would look like:
+
+```ts
+import { sayHelloHandler } from "myapi-typescript-client"; // <- generated typescript client package
```

Comment:
That's a good point! Separating them out would definitely be cleaner. I think it's possible - the tricky bit would be passing the right package names to OpenAPI generator so that we can still use the mustache templates, but import the models from the client package rather than relative paths.

I guess the big pro of this would be separation and you could even switch off generating the handler wrappers, but the con is it'd add another package to build which would slow dev time down a little bit... Would potentially open up the possibility to have eg `typescript-server-lambda` and `typescript-server-express` in future without bloating the clients too much though which could be useful!

Reply:
As discussed offline, we'll rename this to `types` which will encompass both client and server side code. If in future we want to implement different client/server types/wrappers we can add configuration options to let users customise what they want, so as not to bloat the codebase :) This'll be the best of both worlds where the naming is more intuitive, it's still extensible, and users don't take the build time hit of having another package to build :)